src/js/actions/libraries.js

```diff
@@ -499,6 +497,145 @@ define(function (require, exports) {
     };
     renameAsset.reads = [];
     renameAsset.writes = [locks.CC_LIBRARIES, locks.JS_LIBRARIES];
+    
+    /**
+     * Open the specified graphic asset for edit by creating a temp copy of the asset's document and open
+     * in Photoshop. The library store will listen to the document save and close event. 
+     * 
+     * @param {AdobeLibraryElement} element
+     * @return {Promise}
+     */
+    var openGraphicForEdit = function (element) {
+        var representation = element.getPrimaryRepresentation();
+        
+        if (representation.type !== _REP_PHOTOSHOP_TYPE) {
+            // Due to the limitation in DS, we only support editing psd file.
+            log.debug("[CC Lib] unsupported graphic type: " + representation.type);
+            return Promise.resolve();
+        }
+        
+        var tempFilePath,
+            tempPreviewPath;
+        
+        return Promise
+            .bind(this)
+            .then(function () {
+                var graphicEditStatus = this.flux.stores.library.getGraphicEditStatus(element);
+                
+                // If the element already have a temp file, update the temp file path.
+                if (graphicEditStatus) {
+                    tempFilePath = graphicEditStatus.documentPath;
+                    tempPreviewPath = graphicEditStatus.previewPath;
+                    return;
+                }
+                
+                // Otherwise, create a temp copy of the element for edit.
+                return _getTempPaths()
+                    .bind(this)
+                    .then(function (paths) {
```

Comment:
Nesting promises isn't that bad, from waht I've learned... but I don't like this all `then` column here, becauseit does not return anything, and only acts in side effects. So that means paths can be returned to the next then in the chain, and be handled in there synchronously.